File: src/main/java/technology/touchmars/feign/wechat/client/config/WechatErrorDecoder.java
/**
 * 
 */
package technology.touchmars.feign.wechat.client.config;

import java.io.IOException;

import feign.FeignException;
import feign.Response;
import feign.codec.Decoder;
import feign.codec.ErrorDecoder;

/**
 * @author jhcao
 *
 */
public class WechatErrorDecoder implements ErrorDecoder {

	private Decoder decoder;
	
	public WechatErrorDecoder(Decoder decoder) {
		super();
		this.decoder = decoder;
	}

	/* (non-Javadoc)
	 * @see feign.codec.ErrorDecoder#decode(java.lang.String, feign.Response)
	 */
	@Override
	public Exception decode(String methodKey, Response response) {		
		try {
			SimpleResponse errorResponse = 
				(SimpleResponse)decoder.decode(response, SimpleResponse.class);
			return new WechatClientException(errorResponse);
		} catch (FeignException | IOException e) {
			e.printStackTrace();
			return new ErrorDecoder.Default().decode(methodKey, response);
		}		
	}

}

File: src/main/kotlin/technology/touchmars/feign/wechat/client/config/WechatErrorDecoder.kt
/**
 *
 */
package technology.touchmars.feign.wechat.client.config

import feign.FeignException
import feign.Response
import feign.codec.Decoder
import feign.codec.ErrorDecoder
import java.io.IOException

/**
 * @author jhcao
 */
class WechatErrorDecoder(private val decoder: Decoder) : ErrorDecoder {

    /* (non-Javadoc)
	 * @see feign.codec.ErrorDecoder#decode(java.lang.String, feign.Response)
	 */
    @Override
    override fun decode(methodKey: String, response: Response): Exception {
        try {
            val errorResponse = decoder.decode(response, SimpleResponse::class.java) as SimpleResponse
            return WechatClientException(errorResponse)
        } catch (e: FeignException) {
            e.printStackTrace()
            return ErrorDecoder.Default().decode(methodKey, response)
        } catch (e: IOException) {
            e.printStackTrace()
            return ErrorDecoder.Default().decode(methodKey, response)
        }

    }

}
